### ISSUE TITLE
random change of print order on OSX

### ISSUE THREAD

**Issue by rwst** [pynac/pynac#18]
Found in http://trac.sagemath.org/ticket/18155. On Arch and SuSE Linux:
```
sage: -1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
1/2*log(sin(1/8*pi) + 1) - 1/2*log(-sin(1/8*pi) + 1)
```
but apparently on OSX:
```
-1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
```
We would need an OSX account/machine to debug that. Probably a compiler thing, anyway.

**Comment by vbraun** [comment]
I can provide access to the OSX buildbot slave.. just email me your ssh key.

**Comment by rwst** [comment]
I thought I could set up the behaviour without Sage using `ginsh` but (on Linux):
```
> -1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1);
-log(1-sin(1/8*pi))+log(1+sin(1/8*pi))
```
Seems I will have to learn things before really tackling this.

**Comment by vbraun** [comment]
Sage defines its own print order afair

**Comment by rwst** [comment]
Yes, in `order.cpp` but there were no changes there since pynac-0.3.2.

**Comment by behackl** [comment]
As I'm not really an expert for C++ code, this probably is nothing more than a (more or less) educated guess: is it possible that the changes in 544309901c51c19fdfd3e9920e2a51d70ff94f9b somehow mess up the order of the summands (at least on some platforms...)?

**Comment by rwst** [comment]
It's actually random:
```
┌───────────────────────────────────────────────────────────┐
│ SageMath Version 6.7.beta1, Release Date: 2015-04-15               │
....
sage: h(x)=sec(x)
sage: integral(h,(x,0,pi/8))
1/2*log(sin(1/8*pi) + 1) - 1/2*log(-sin(1/8*pi) + 1)
sage: integral(sec(x),(x,0,pi/8))
-1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
sage: integral(h,(x,0,pi/8))
-1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
```
This is about 50-50. And:
┌───────────────────────────────────────────────────────────┐
│ SageMath Version 6.7.beta1, Release Date: 2015-04-15               │
....
sage: -1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
1/2*log(sin(1/8*pi) + 1) - 1/2*log(-sin(1/8*pi) + 1)
sage: -1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
1/2*log(sin(1/8*pi) + 1) - 1/2*log(-sin(1/8*pi) + 1)
sage: -1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
-1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
sage: -1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
1/2*log(sin(1/8*pi) + 1) - 1/2*log(-sin(1/8*pi) + 1)
sage: -1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
1/2*log(sin(1/8*pi) + 1) - 1/2*log(-sin(1/8*pi) + 1)
sage: -1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
-1/2*log(-sin(1/8*pi) + 1) + 1/2*log(sin(1/8*pi) + 1)
```
And now the next positive start appears at the 17th try, so it's not 50-50.
On Linux both are positive 20 out of 20 tries.

**Comment by rwst** [comment]
Also, the reduced `-1/2*log(-sin(1/8*pi)) + 1/2*log(sin(1/8*pi))` yields the same on both systems, so the question comes up what is in fact the correct answer to expect for the original input.

**Comment by behackl** [comment]
Alright. Do you have an idea where this random behavior might come from?

Also, what do you mean with the "reduced" expression above? `log(-sin(1/8*pi))` doesn't really make that much sense to me...

**Comment by rwst** [comment]
It was introduced with d2d5ca07dd2ed7bdd37d6c, so we need no longer speculate. Maybe it's best to just revert it...

**Comment by vbraun** [comment]
I guess the caching is what kills us. We use different orders for computation and printing, but if you cache the ordering then thats not going to work so hot.

**Comment by behackl** [comment]
It makes sense that this comes from something like caching. +1 for reverting (as long as the doctests pass -- and especially the behavior of `integral(sec(x), (x, 0, pi/8))` needs to stabilize on OSX ;-)).

**Comment by rwst** [comment]
Tests show nothing beyond what would have been fixed in #18155, so it's green. Let's do a quick release.
